### ISSUE TITLE
Seemingly incorrect resolving API endpoint IP

### ISSUE THREAD

**Issue by emmetog** [vmware-tanzu/sonobuoy#1162]
**What steps did you take and what happened:**
I've set up a k8s cluster with 3 masters and 1 worker with 2 load balancers sharing a virtual IP (10.0.0.201) in front of the API service. I've added a line in `/etc/hosts` to map the hostname (`k8s-cluster-endpoint`) to the virtual IP (10.0.0.201). From what I can see everything is working correctly, all the following pods seem to be fine and kubectl which is pointing to the virtual IP is working fine.

```
root@master01:sonobuoy# kubectl get pods --all-namespaces
NAMESPACE     NAME                                           READY   STATUS    RESTARTS   AGE
kube-system   coredns-66bff467f8-8s6nl                       1/1     Running   28         5h58m
kube-system   coredns-66bff467f8-tnw2c                       1/1     Running   28         5h58m
kube-system   etcd-master01                                  1/1     Running   1          6h1m
kube-system   etcd-master02                                  1/1     Running   4          4h2m
kube-system   etcd-master03                                  1/1     Running   1          4h1m
kube-system   kube-apiserver-master01                        1/1     Running   1          6h1m
kube-system   kube-apiserver-master02                        1/1     Running   4          4h3m
kube-system   kube-apiserver-master03                        1/1     Running   1          4h1m
kube-system   kube-controller-manager-master01               1/1     Running   3          6h1m
kube-system   kube-controller-manager-master02               1/1     Running   4          4h3m
kube-system   kube-controller-manager-master03               1/1     Running   1          4h1m
kube-system   kube-proxy-mhmvx                               1/1     Running   1          5h58m
kube-system   kube-proxy-n7k9w                               1/1     Running   1          4h1m
kube-system   kube-proxy-w4m5z                               1/1     Running   4          4h3m
kube-system   kube-proxy-wtncr                               1/1     Running   3          4h1m
kube-system   kube-router-j8mpv                              1/1     Running   0          22m
kube-system   kube-router-pdfzm                              1/1     Running   0          22m
kube-system   kube-router-s7n79                              1/1     Running   0          22m
kube-system   kube-router-zz2t4                              1/1     Running   0          23m
kube-system   kube-scheduler-master01                        1/1     Running   2          6h1m
kube-system   kube-scheduler-master02                        1/1     Running   4          4h3m
kube-system   kube-scheduler-master03                        1/1     Running   1          4h1m

```
However when running sonobuoy it seems to be trying to use 10.96.0.1:443 as the api server ip, despite the fact that the server endpoint is `https://k8s-cluster-endpoint:6443`. I'm unsure where `10.96.0.1:443` is coming from.
```
root@master01:sonobuoy# cat /etc/kubernetes/admin.conf | grep endp
    server: https://k8s-cluster-endpoint:6443
```

Here's how I run sonobuoy:
```
$ sonobuoy --kubeconfig /etc/kubernetes/admin.conf run --mode quick --wait
```
And here are the logs:
```
root@master01:root# kubectl logs --follow -n sonobuoy sonobuoy
time="2020-08-27T21:30:43Z" level=info msg="Scanning plugins in ./plugins.d (pwd: /)"
time="2020-08-27T21:30:43Z" level=info msg="Scanning plugins in /etc/sonobuoy/plugins.d (pwd: /)"
time="2020-08-27T21:30:43Z" level=info msg="Directory (/etc/sonobuoy/plugins.d) does not exist"
time="2020-08-27T21:30:43Z" level=info msg="Scanning plugins in ~/sonobuoy/plugins.d (pwd: /)"
time="2020-08-27T21:30:43Z" level=info msg="Directory (~/sonobuoy/plugins.d) does not exist"
time="2020-08-27T21:31:13Z" level=error msg="could not get api group resources: Get \"https://10.96.0.1:443/api?timeout=32s\": dial tcp 10.96.0.1:443: i/o timeout"
time="2020-08-27T21:31:13Z" level=info msg="no-exit was specified, sonobuoy is now blocking"
```

**What did you expect to happen:**
The sonobuoy run to be able to connect to the API to get the resources.

**Anything else you would like to add:**
Here is my `/etc/hosts` on `master01` (other nodes are similar, only the localhost name changes):
```
127.0.0.1   localhost master01

10.0.0.201         k8s-cluster-endpoint
10.0.0.32          ha01
10.0.0.12          ha02
10.0.0.33          master01
10.0.0.15          master02
10.0.0.30          master03
10.0.0.34          worker01

# The following lines are desirable for IPv6 capable hosts
::1     localhost ip6-localhost ip6-loopback
ff02::1 ip6-allnodes
ff02::2 ip6-allrouters
```

**Environment:**

- Sonobuoy version: 
```
root@master01:sonobuoy# sonobuoy version
Sonobuoy Version: v0.18.5
MinimumKubeVersion: 1.16.0
MaximumKubeVersion: 1.18.99
GitSHA: a79a43a99680f7274e6315dbf09fadc8fd81a6e0
API Version check skipped due to missing `--kubeconfig` or other error
```
- Kubernetes version: (use `kubectl version`):
```
root@master01:sonobuoy# kubectl version
Client Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.8", GitCommit:"9f2892aab98fe339f3bd70e3c470144299398ace", GitTreeState:"clean", BuildDate:"2020-08-13T16:12:48Z", GoVersion:"go1.13.15", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.0", GitCommit:"9e991415386e4cf155a24b1da15becaa390438d8", GitTreeState:"clean", BuildDate:"2020-03-25T14:50:46Z", GoVersion:"go1.13.8", Compiler:"gc", Platform:"linux/amd64"}
```

- Kubernetes installer & version:
```
root@master01:sonobuoy# kubeadm version
kubeadm version: &version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.8", GitCommit:"9f2892aab98fe339f3bd70e3c470144299398ace", GitTreeState:"clean", BuildDate:"2020-08-13T16:10:16Z", GoVersion:"go1.13.15", Compiler:"gc", Platform:"linux/amd64"}
```
I've created the cluster using kubeadm like so:
```
kubeadm init --config /etc/kubernetes/init/kubeadm-init.yaml --ignore-preflight-errors=NumCPU
```
where `/etc/kubernetes/init/kubeadm-init.yaml` is:
```yml
apiVersion: kubeadm.k8s.io/v1beta2
kind: ClusterConfiguration
kubernetesVersion: v1.18.0
clusterName: test-int

controlPlaneEndpoint: "k8s-cluster-endpoint"

controllerManager:
  extraArgs:
    allocate-node-cidrs: "true"
    cluster-cidr: "10.244.0.0/16"

apiServer:
  extraArgs:
    enable-admission-plugins: "NamespaceLifecycle,LimitRanger,ServiceAccount,DefaultStorageClass,DefaultTolerationSeconds,MutatingAdmissionWebhook,ValidatingAdmissionWebhook,Priority,ResourceQuota,PodSecurityPolicy"
```

- Cloud provider or hardware configuration:

I'm using kube-router, the configuration is as follows:
```yml
apiVersion: v1
kind: ConfigMap
metadata:
  name: kube-router-cfg
  namespace: kube-system
  labels:
    tier: node
    k8s-app: kube-router
data:
  cni-conf.json: |
    {
       "cniVersion":"0.4.0",
       "name":"mynet",
       "plugins":[
          {
             "name":"kubernetes",
             "type":"bridge",
             "bridge":"kube-bridge",
             "isDefaultGateway":true,
             "ipam":{
                "type":"host-local"
             }
          }
       ]
    }
---
apiVersion: apps/v1
kind: DaemonSet
metadata:
  name: kube-router
  namespace: kube-system
  labels:
    k8s-app: kube-router
    tier: node
spec:
  selector:
    matchLabels:
      k8s-app: kube-router
      tier: node
  template:
    metadata:
      labels:
        k8s-app: kube-router
        tier: node
    spec:
      priorityClassName: system-node-critical
      serviceAccountName: kube-router
      serviceAccount: kube-router
      containers:
        - name: kube-router
          image: docker.io/cloudnativelabs/kube-router
          imagePullPolicy: Always
          args:
            - --run-router=true
            - --run-firewall=true
            - --run-service-proxy=true
            - --bgp-graceful-restart=true
            - --kubeconfig=/var/lib/kube-router/kubeconfig
          env:
            - name: NODE_NAME
              valueFrom:
                fieldRef:
                  fieldPath: spec.nodeName
            - name: KUBE_ROUTER_CNI_CONF_FILE
              value: /etc/cni/net.d/10-kuberouter.conflist
          livenessProbe:
            httpGet:
              path: /healthz
              port: 20244
            initialDelaySeconds: 10
            periodSeconds: 3
          resources:
            requests:
              cpu: 250m
              memory: 250Mi
          securityContext:
            privileged: true
          volumeMounts:
            - name: lib-modules
              mountPath: /lib/modules
              readOnly: true
            - name: cni-conf-dir
              mountPath: /etc/cni/net.d
            - name: kubeconfig
              mountPath: /var/lib/kube-router
              readOnly: true
            - name: xtables-lock
              mountPath: /run/xtables.lock
              readOnly: false
      initContainers:
        - name: install-cni
          image: docker.io/cloudnativelabs/kube-router
          imagePullPolicy: Always
          command:
            - /bin/sh
            - -c
            - set -e -x;
              if [ ! -f /etc/cni/net.d/10-kuberouter.conflist ]; then
              if [ -f /etc/cni/net.d/*.conf ]; then
              rm -f /etc/cni/net.d/*.conf;
              fi;
              TMP=/etc/cni/net.d/.tmp-kuberouter-cfg;
              cp /etc/kube-router/cni-conf.json ${TMP};
              mv ${TMP} /etc/cni/net.d/10-kuberouter.conflist;
              fi
          volumeMounts:
            - name: cni-conf-dir
              mountPath: /etc/cni/net.d
            - name: kube-router-cfg
              mountPath: /etc/kube-router
      hostNetwork: true
      tolerations:
        - effect: NoSchedule
          operator: Exists
        - key: CriticalAddonsOnly
          operator: Exists
        - effect: NoExecute
          operator: Exists
      volumes:
        - name: lib-modules
          hostPath:
            path: /lib/modules
        - name: cni-conf-dir
          hostPath:
            path: /etc/cni/net.d
        - name: kube-router-cfg
          configMap:
            name: kube-router-cfg
        - name: kubeconfig
          configMap:
            name: kube-proxy
            items:
              - key: kubeconfig.conf
                path: kubeconfig
        - name: xtables-lock
          hostPath:
            path: /run/xtables.lock
            type: FileOrCreate
---
apiVersion: v1
kind: ServiceAccount
metadata:
  name: kube-router
  namespace: kube-system
---
kind: ClusterRole
apiVersion: rbac.authorization.k8s.io/v1beta1
metadata:
  name: kube-router
  namespace: kube-system
rules:
  - apiGroups:
      - ""
    resources:
      - namespaces
      - pods
      - services
      - nodes
      - endpoints
    verbs:
      - list
      - get
      - watch
  - apiGroups:
      - "networking.k8s.io"
    resources:
      - networkpolicies
    verbs:
      - list
      - get
      - watch
  - apiGroups:
      - extensions
    resources:
      - networkpolicies
    verbs:
      - get
      - list
      - watch
---
kind: ClusterRoleBinding
apiVersion: rbac.authorization.k8s.io/v1beta1
metadata:
  name: kube-router
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: kube-router
subjects:
  - kind: ServiceAccount
    name: kube-router
    namespace: kube-system
```

- OS (e.g. from `/etc/os-release`): 
```
root@master01:sonobuoy# cat /etc/os-release
NAME="Ubuntu"
VERSION="18.04.5 LTS (Bionic Beaver)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 18.04.5 LTS"
VERSION_ID="18.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=bionic
UBUNTU_CODENAME=bionic
```
- Sonobuoy tarball (which contains * below)
https://github.com/vmware-tanzu/sonobuoy/releases/download/v0.18.5/sonobuoy_0.18.5_linux_amd64.tar.gz

**Comment by emmetog** [comment]
It looks like this is not an issue with sonobuoy, it's also happening when I try to deploy ingress-nginx. I suspect that it may be something wrong with my kube-router or coredns setup, but I'm still not sure what's wrong.

Feel free to close this as not sonobuoy related, although any pointers towards a fix would be much appreciated.

**Comment by vladimirvivien** [comment]
Thanks for reaching out @emmetog. Good luck fixing your issue.

**Comment by emmetog** [comment]
For future reference, the issue was kube-router which wasn't set up correctly. As a K8s noobie a few things were tripping me up:
* In my attempts to debug and fix this, I was changing all sorts of parameters to `kubeadm init` such as cluster-cidr, service-cluster-ip-range, advertise-address. In the end none of these are needed, kube-router just requires `allocate-node-cidrs: "true"`. When this wasn't correctly configured I was getting errors in kube-router like `Failed to list *v1.Endpoints: Get https://10.96.0.1:443/api/v1/endpoints?limit=500&resourceVersion=0: dial tcp 10.96.0.1:443: i/o timeout`
* I had also enabled some extra admission plugins, such as `PodSecurityPolicy` which restrict inter-pod communication. When I found it in the docs during my blind stumbling it sounded so great that I couldn't help but turn it on. Lesson: get everything working with a basic setup before turning on things like PSP.
* Giving kube-router permissions to reach the api. I had to investigate quite a bit before realizing that the kube-router pod doesn't actually need a kubeconfig file (most examples I had seen online had lots of mentions of the kubeconfig file in the DaemonSet manifests). Turns out this is completely unnecessary as k8s automatically injects the tokens into each pod, so the kube-router already had all it needed to contact the k8s api. Commenting out all mention of kubeconfig in the DaemonSet manifest for kube-router fixed `403 permission denied` issues.
